src/time/location.cr

```diff
@@ -0,0 +1,281 @@
+  class InvalidLocationNameError < Exception
+    getter name, source
+
+    def initialize(@name : String, @source : String? = nil)
+      super "Invalid location name: #{name}"
+    end
+  end
+
+  class InvalidTimezoneOffsetError < Exception
+    def initialize(offset : Int)
+      super "Invalid time zone offset: #{offset}"
+    end
+  end
+
+  struct Zone
+    UTC = new "UTC", 0, false
+
+    getter name : String
+    getter offset : Int32
+    getter? dst : Bool
+
+    def initialize(@name : String, @offset : Int32, @dst : Bool)
+      # Maximium offets of IANA timezone database are -12:00 and +14:00.
+      # +/-24 hours allows a generous padding for unexpected offsets.
+      # TODO: Maybe reduce to Int16 (+/- 18 hours).
+      raise InvalidTimezoneOffsetError.new(offset) if offset >= SECONDS_PER_DAY || offset <= -SECONDS_PER_DAY
+    end
+
+    def inspect(io : IO)
+      io << "Time::Zone<"
+      io << offset
+      io << ", " << name
+      io << " (DST)" if dst?
+      io << '>'
+    end
+  end
+
+  # :nodoc:
+  record ZoneTransition, when : Int64, index : UInt8, standard : Bool, utc : Bool do
+    getter? standard, utc
+
+    def inspect(io : IO)
+      io << "Time::ZoneTransition<"
+      io << '#' << index << ", "
+      Time.epoch(self.when).to_s("%F %T", io)
+      io << ", STD" if standard?
+      io << ", UTC" if utc?
+      io << '>'
+    end
+  end
+
+  # Describes the Coordinated Universal Time (UTC).
+  UTC = new "UTC", [Zone::UTC]
+
+  property name : String
+  property zones : Array(Zone)
+
+  # Most lookups will be for the current time.
+  # To avoid the binary search through tx, keep a
+  # static one-element cache that gives the correct
+  # zone for the time when the Location was created.
+  # The units for @cached_range are seconds
+  # since January 1, 1970 UTC, to match the argument
+  # to `#lookup`.
+  @cached_range : Tuple(Int64, Int64)
+  @cached_zone : Zone
+
+  # Creates a `Location` instance named *name* with fixed *offset*.
+  def self.fixed(name : String, offset : Int32)
+    new name, [Zone.new(name, offset, false)]
+  end
+
+  # Creates a `Location` instance with fixed *offset*.
+  def self.fixed(offset : Int32)
+    span = offset.abs.seconds
+    name = sprintf("%s%02d:%02d", offset.sign < 0 ? '-' : '+', span.hours, span.minutes)
+    fixed name, offset
+  end
+
+  # Returns the `Location` with the given name.
+  #
+  # This uses an array of paths to look for timezone data. Each path can
+  # either point to a directory or an uncompressed ZIP file.
+  #
+  # The first timezone data matching the given name that is successfully loaded
+  # and parsed is returned.
+  #
+  # Special names:
+  # * `UTC` and empty string return `Location::UTC`
+  # * `Local` returns `Location.local`
+  #
+  # An custom lookup location can be added as environment variable `ZONEFILE`.
+  #
+  # Example:
+  # `ZONEFILE=/path/to/zoneinfo.zip crystal eval 'pp Location.load("Custom/Location")'`
+  def self.load(name : String)
```

Comment:
I would make this return `nil` if the location is not found. Raising an exception is expensive.

Reply:
But then you can't just simply do `Time.new(..., location: Location.load("Europe/Berlin"))`.

Reply:
I think it should be provided that a location can always be loaded (as long as it is a valid name).

Reply:
Right. Well, we could then have `Location.load?` that returns `nil`.

I actually thought this would be needed when parsing a timezone, but now I think it has nothing to do. So for now we can leave it like that (we can always add `load?` later if needed).

Reply:
There is also `.load?` (undocumented) which is fast without an exception, allthough it expects `zone_sources` as second argument. But this could have a default value of `Crystal::System::Time.zone_sources`.